충남대 FE 정지원 - 1단계 React Query 적용

jjw5655 opened the pull request:
안녕하세요 멘토님, 1단계 구현입니다.
커밋메시지 참고 부탁드립니다. 감사합니다!

jjw5655 committed:
docs: 미션 4 - 1단계 과제 요구사항으로 README.md 업데이트

jjw5655 committed:
chore: React Query 기본 설정 및 보안 취약점 일부 수정

- npm install @tanstack/react-query 명령어로 패키지 설치
- QueryClient, QueryClientProvider를 앱 최상위(main.tsx)에 적용하여 React Query를 사용할 수 있도록 설정
- npm audit fix 명령어로 의존성 보안 취약점 일부 자동 수정

jjw5655 committed:
feat: useInfiniteScroll 훅을 hooks 배럴 파일에 export로 추가

jjw5655 committed:
feat: 상품/랭킹/테마 데이터 패칭(GET)에 React Query useQuery 적용

- useProduct, useRankingProducts, useThemes 훅을 useQuery 기반으로 리팩터링
- 기존 useFetch 의존성 제거
- useProduct 훅 반환값 변경에 따라 OrderPage에서 { product } → { data: product }로 destructuring 수정

jjw5655 committed:
feat: 주문 생성(POST)에 React Query useMutation 적용

- OrderPage에서 postOrder 직접 호출 대신 useMutation으로 주문 생성 처리
- 성공/실패 후처리를 onSuccess, onError 옵션으로 선언적으로 관리

jjw5655 committed:
feat: Suspense 및 ErrorBoundary를 통한 선언적 데이터 로딩/에러 처리 적용

- 공통 ErrorBoundary 컴포넌트 추가
- OrderPage에 Suspense와 ErrorBoundary 적용
- Suspense fallback에 기존 로딩 UI 사용

jjw5655 committed:
refactor: 불필요한 useFetch 훅 제거

- useFetch 훅 및 관련 코드/의존성 제거
- hooks/index.ts에서 useFetch export 제거

terajh commented on src/main.tsx:
```diff
@@ -6,15 +6,20 @@ import './index.css';
 import { AuthProvider } from '@/contexts';
+import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
+
+const queryClient = new QueryClient();
```
[참고의견!] QueryClient 생성 시 suspense, staleTime, retry 등 기본 옵션을 설정해두면 각 훅에서 반복적으로 설정할 필요가 없어요! 알고계실수도 있지만 혹시나 추후 공통적으로 사용이 필요한 설정이 있다면 사용해보면 좋을 것 같아 의견 남겨보아요

```ts
const queryClient = new QueryClient({
  defaultOptions: {
    queries: { suspense: true, retry: 1, staleTime: 1000 * 60 },
  },
});
```

jjw5655 replied:
아니요 몰랐습니다. 전혀 몰랐어요. 이렇게 잘 모르고 쓰는 경우가 많아서, 이런부분 집어주시면 공부에 많은 도움이 되는거같습니다. 감사합니다!

옵션 중 suspense부분이 오류메시지가 떠서 확인해보니 react query v4부터 suspense의 전역설정이 불가능하도록 바뀌었다고 하네요. 따라서 useQuery 훅에서 suspense: true로 개별 설정하거나 useSuspenseQuery 훅을 사용해야 할 것 같습니다.

terajh replied:
오 그렇군요! 덕분에 v4부터 suspense의 전역설정이 불가능하단점을 알게되었네요 감사합니다ㅎㅎ

terajh commented on src/components/common/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,53 @@
+            <button onClick={this.handleReset} style={{ marginTop: 16 }}>
+              다시 시도
+            </button>
+          </div>
```
[소소의견] 에러 발생시 보여주는 fallbackComponent 의 경우 따로 컴포넌트로 분리해주면 이후 다른곳에서 자유롭게 사용할 수 있을 것 같아 의견 남겨봅니다!

jjw5655 replied:
별도 컴포넌트로 분리했습니다.

terajh commented on src/pages/OrderPage.tsx:
```diff
@@ -263,13 +266,36 @@ function handleOrderError(
+        totalPrice
+      );
+      alert(msg);
+      navigate('/');
```
[소소의견] 아래 handleOrderError 로 나눠주신것처럼 onSuccessCreateOrder 와 같은 콜백함수를 따로 선언해주면 코드 읽기가 더 편해질 것 같네요ㅎㅎ (지금도 좋습니다~)

jjw5655 replied:
다른 함수와 같이 통일성있게 분리했습니다!

terajh commented on src/pages/OrderPage.tsx:
```diff
@@ -263,13 +266,36 @@ function handleOrderError(
+  const { data: product, isLoading, error } = useProduct(productId ?? '');
+
+  // 주문 생성 useMutation
+  const orderMutation = useMutation({
```
[질문] 다른 파일에서 useQuery를 활용해 API 호출 로직을 잘 분리해주신 부분이 눈에 띄었습니다. 이와 마찬가지로 전체적으로 사용되는 useQuery/useMutation 기반 API 요청 로직들을 한곳에 모아 관리해보는 방법을 사용할 수 있을 것 같아요. 이렇게 하면 api 에 대한 공통 처리(에러 핸들링, 인터셉터 추가 등)를 한 곳에서 유지보수할 수 있고, API 스펙 변경 시에도 수정 범위를 최소화할 수 있을 것 같아요

예를 들어, src/api 폴더를 두고 그 하위에
```
theme/query.ts
theme/mutations.ts
```
처럼 도메인별로 query/mutation을 나누어 관리하는 방식도 고려해볼 수 있을 것 같아요. 이런 구조에 대해 고민해보시면 좋은 경험이 되실것 같아 어떻게 생각하시는지 질문남겨봅니다 ㅎ

jjw5655 replied:
api들이 각각 목표에 맞게 특화되어 설정되어있고, 그에 따라 유즈쿼리방법도 정해지니 제안해주신 방법대로 도메인마다 분리하는게 좋아보입니다. 하나의 api폴더에 api관련된 모든걸 박아두는게 추후 유지보수나 확장에서 이점이 있을 것 같아요.

도메인별로 api함수들과 react query 훅을 분리하여 리펙토링 했습니다. API명세를 보고, theme, product뿐만 아니라 login과 order도 통일성있게 분리했습니다.
> 기존: 
> - api폴더 내에 도메인별 폴더를 나누어져 있지 않음
> - API 관련 React Query 훅들이 hook폴더에 분산되어 있음.
> - 일부 파일은 useMutation을 직접 import하여 사용함. import { useMutation } from '@tanstack/react-query';
> 
> 수정:
> - api 폴더 내에 도메인별로 폴더를 생성하여 관련 API 함수와 React Query 훅(query/mutations)을 한 곳에 모아 관리
> - API 관련 훅은 모두 도메인별 api 폴더로 이동, hooks 폴더에는 비API(로컬/UX) 훅만 남김
> - useMutation 등 React Query 훅도 도메인별 커스텀 훅으로 래핑하여 직접 import하지 않도록 일원화

엮여있는 파일들이 많다보니, 이렇게 뒤집어 엎어서 리펙토링하는게 생각보다 시간이 많이 걸리네요.. 
보통 API 명세가 주어지면, 이런식으로 API폴더에 관련 함수/훅들을 모두 만들어두고 사용하는게 일반적인가요?

terajh replied:
많은 고민을 통해 개선해 주신 부분이 잘 느껴졌습니다. 고생 많으셨습니다.

“보통 API 명세가 주어지면 이렇게 API 폴더에 관련 함수와 훅을 모두 만들어 두고 사용하는 게 일반적인가요?”라는 질문을 주셨는데, 팀마다 개발 방향은 다를 수 있지만, 관심사 분리를 위해 별도의 계층을 두는 경우가 많습니다.

API 명세는 대부분 서버에서 제공되며, 언제든 수정될 수 있는 가변적인 요소라고 볼 수 있습니다. 그렇기 때문에 이후 변경 시에도 쉽게 대응할 수 있도록 유지보수를 고려한 구조를 설계하는 것이 중요하다고 생각합니다. 그 방법 중 하나가 바로 API 호출 로직을 별도의 파일로 분리하는 것일 것 같은데요.

다만, 도메인이 많아질수록 지나치게 세분화된 분리는 오히려 파일 수를 급격히 늘려 유지보수를 어렵게 할 수도 있습니다. 이런 경우에는
- api/ 디렉토리 하위에 도메인별 파일을 배치하거나,
- api/query.ts, api/mutation.ts와 같이 호출 성격에 따라 통합 관리하는 방식
- 등등등

위의 여러 해결방법들을 같이 개발하는 동료와 논의후 균형을 맞추는 방법을 고려할 수 있을 것 같습니다. 정답은 없으니까요ㅎㅎ

terajh reviewed (changes requested):
지원님~ 1단계 미션 고생 많으셨습니다!

전반적으로 useQuery 사용 등 요구사항에 맞춰 코드를 잘 작성해주셨고, 구조적인 측면에서도 깔끔하게 정리해주신 점이 인상 깊었습니다. 커밋 단위도 신경 써서 잘 나눠주신 게 느껴졌어요. 👍 useQuery 에 대한 사용도 깔끔하게 작업해주신 것 같구요. 추가로 참고만 할법한 내용들을 코멘트로 남겨두었고, 질문도 하나 추가해두어 PR은 request changes 를 드려보겠습니다! 확인후 다시 리뷰요청 부탁드려요!

과제를 진행하시면서 고민되었던 부분이나 궁금한 점이 있다면, PR description이나 DM으로 언제든 편하게 남겨주세요. 감사합니다! 🙏🏻

jjw5655 committed:
feat: QueryClient 생성 시 기본값 설정

- QueryClient의 defaultOptions에 queries 옵션 추가
  - retry: 1 (1회 재시도)
  - staleTime: 1분(1000 * 60ms)
- suspense 옵션은 React Query v4부터 defaultOptions에서 전역 설정이 불가하여 주석 처리함
- 각 쿼리 훅에서 반복적으로 옵션을 지정하지 않아도 되도록 개선

jjw5655 committed:
refactor: ErrorBoundary의 fallback UI를 ErrorFallback 컴포넌트로 분리

- 에러 발생 시 보여주는 fallback UI를 별도의 ErrorFallback 컴포넌트로 분리하여 재사용성 향상
- ErrorBoundary에서 ErrorFallback을 사용하도록 수정
- error prop의 타입 일치 문제(null → undefined) 해결

jjw5655 committed:
refactor: 주문 성공 처리 로직을 onSuccessCreateOrder 함수로 분리

- 주문 성공 시 안내 메시지 및 이동 로직을 onSuccessCreateOrder 함수로 분리하여 가독성 및 일관성 향상
- 기존 handleOrderError와 구조적으로 대칭되도록 개선

jjw5655 committed:
refactor: theme 도메인 API/쿼리 분리

- theme 관련 API 함수(getThemes, getThemeInfo, getThemeProducts) src/api/theme/apis.ts로 이동
- theme 관련 useQuery 기반 쿼리 훅을 src/api/theme/query.ts로 분리
- 배럴 파일(index.ts) 구조 적용
- theme 도메인 분리에 맞게 기존의 '@/api/themes' import 경로를 '@/api/theme'로 일괄 변경

jjw5655 committed:
refactor: product 도메인 API/쿼리 분리

- product 관련 API 함수(getRankingProducts, getProductById, getProductSummary) src/api/product/apis.ts로 이동
- product 관련 useQuery 기반 쿼리 훅(useProduct, useRankingProducts) src/api/product/query.ts로 분리
- 배럴 파일(index.ts) 구조 적용
- 기존 훅(useProduct, useRankingProducts)은 내부적으로 새 구조를 사용하도록 변경

jjw5655 committed:
refactor: order 도메인 API/뮤테이션 분리

- 주문 생성 API(postOrder) 함수를 src/api/order/apis.ts로 이동
- 주문 생성용 useMutation 커스텀 훅(useOrderMutation) src/api/order/mutations.ts로 분리
- 배럴 파일(index.ts) 구조 적용
- OrderPage에서 useOrderMutation 훅을 사용하도록 리팩토링 및 타입 에러 해결

jjw5655 committed:
chore: product API 명세 기반 미구현 엔드포인트 TODO 주석 추가

- API 명세서를 참고하여 아래 3개 엔드포인트의 미구현 상태를 src/api/product/apis.ts 상단에 TODO로 표기해두었음.
  - /api/products/:productId/detail (상세 정보)
  - /api/products/:productId/wish (찜 정보)
  - /api/products/:productId/highlight-review (리뷰)

jjw5655 committed:
refactor: auth 도메인 API 분리

- 로그인 API(loginApi) 함수를 src/api/auth/apis.ts로 이동
- 배럴 파일(index.ts) 구조 적용
- AuthContext에서 loginApi import 경로를 '@/api/auth'로 변경

jjw5655 committed:
refactor: api/index.ts를 공통 apiClient 인스턴스만 남기고 정리

- 도메인별 API 함수(export 등) 제거
- 공통 axios 인스턴스(apiClient) 생성 및 export만 유지
- 각 도메인별 API 파일에서 공통 인스턴스만 import해서 사용하도록 구조 일원화

jjw5655 committed:
refactor: API 관련 훅 파일 및 export 정리

- src/hooks 내 useThemes, useProduct, useRankingProducts 파일 삭제
- src/hooks/index.ts에서 해당 훅들의 export 제거
- API 관련 React Query 훅은 도메인별 src/api/폴더에서만 관리하도록 구조 일원화

jjw5655 committed:
refactor: API 관련 훅 import 경로를 도메인별 api 폴더로 일괄 변경

- useThemes → useThemesQuery 및 import 경로를 '@/api/theme'로 변경
- useRankingProducts, useProduct, ProductSummary import 경로를 '@/api/product'로 변경
- 삭제된 훅 파일로 인한 import 에러 해결

terajh reviewed (approved):
지원님께서 드린 코멘트를 적극적으로 고민하시고 코드까지 리팩터링해 주셨네요. 고생 많으셨습니다 👍

서비스 전반에서 사용되는 API 호출 패턴은 언제나 중요한 고민 포인트라고 생각합니다. 변경에 유연하게 대응할 수 있도록, 그리고 쉽게 활용할 수 있도록 구조를 다듬어 나가는 과정은 항상 쉽지 않지만, 이번에 깊이 고민해보신 경험은 앞으로 큰 자산이 될 거라고 생각합니다.

수정사항들은 확인되어서 PR은 승인 후 머지하도록 하겠습니다. 코멘트로 남겨주신 질문에 대해서도 답변을 달아두었으니 확인해 보시고 지원님만의 기준을 세우는 데 참고부탁드립니다 😊
